### DISCUSSION TITLE
Can a downstream model trained on Clay embeddings from sensor A operate on Clay embeddings generated from sensor B?

### DISCUSSION THREAD

**Discussion by rsousa95** [Clay-foundation/model#385] [Q&A]
Following the tutorials, I've managed to train a segmentation model for fire detection. I've used an open source dataset with Sentinel-2 data, I froze the encoder and trained the segmentation head. The results I got are quite good. 

Now, how feasible is it to apply this trained model with images from another sensor? Since the segmentation head works on the embeddings from the encoder, and the embeddings are somewhat model agnostic, ideally I should be able to use this trained model on the new sensor. The embedding space is shared by multiple sensors, right?

I want to use it for NAIP images. I've tried applying it, feeding the model the datacube with the changed means and std for normalization, the GSD and the wavelengths. So far I'm not getting anything at all but maybe I'm just normalizing wrong.

**Comment by Yigtwxx** [comment]
The shapes line up, which is why you get silence instead of an error. Whether the *space* lines up is a weaker claim than it looks, and it's weakest in exactly the part a segmentation head uses.

Band count never reaches the transformer. The patch embedding is the DOFA-style hypernetwork in `claymodel/factory.py`: `waves` goes through `WavesTransformer`, which emits the conv weights, and those get rearranged to `cout cin k1 k2`. Ten S2 bands and four NAIP bands both collapse to `L = (H/patch) * (W/patch)` tokens of dim `D`. Your head will happily consume either.

On "the embedding space is shared" — look at what the pretraining objective actually ties together (`claymodel/model.py`, `ClayMAE.forward`):

```python
representations = self.proj(encoded_unmasked_patches[:, 0, :])  # [B D']
...
indices = self.metadata[platform].rgb_indices
rgb = datacube["pixels"][:, indices, :, :]
target = self.teacher(self.teacher_resize(rgb))
representation_loss = 1.0 - F.cosine_similarity(representations, target).mean()
loss = 0.9 * reconstruction_loss + 0.1 * representation_loss
```

The only cross-sensor pull is that cosine term. It's weighted 0.1, it goes through a teacher that sees **RGB only**, and it acts on `[:, 0, :]` — the class token. The patch tokens are shaped by per-platform masked reconstruction, and nothing in the loss asks a NAIP patch token and an S2 patch token of the same ground to land in the same place. A classification head on the CLS embedding has some reason to transfer. A segmentation head on patch tokens has much less.

Three things I'd check, in order.

**Band order.** These two are not in the same order in `configs/metadata.yaml`:

```yaml
sentinel-2-l2a:
  band_order: [blue, green, red, rededge1, ..., swir16, swir22]
naip:
  band_order: [red, green, blue, nir]
```

RGB is reversed between them. Since the hypernetwork generates one conv kernel per entry of `waves`, the order of `waves` has to match the channel axis of your cube exactly. Get it wrong and nothing raises — you just get a well-formed embedding of the wrong thing. Same for the statistics: NAIP means are `110.16 / 115.41 / 98.15 / 139.04` on an 8-bit DN scale, S2 means are `1105 / 1355 / 1552 / ...` on the scaled-reflectance one. If any S2 normalization survived into your NAIP path the input is off by an order of magnitude before the encoder sees it.

**GSD does more than you'd expect.** It's not metadata that rides along, it rescales the positional encoding directly (`claymodel/utils.py`):

```python
omega = torch.arange(dim // 4) / (dim // 4 - 1)
omega = 1.0 / (temperature ** (2 * omega / dim)) * (gsd / 1.0)
```

`gsd=10` gives frequencies 10x those of `gsd=1.0`, and this is added straight onto the patch tokens before the transformer. Your head was trained on tokens carrying the S2 version. Physically it's the right thing to do — position is being expressed in ground units — but it means a 256-pixel NAIP chip spans 256 units where a 256-pixel S2 chip spans 2560.

That gives you a cheap test: resample your NAIP imagery to 10 m and pass `gsd=10`, so the chip covers the same ground area at the same scale the head was trained on. You throw away all the resolution you were hoping to gain, but if it starts producing sensible masks you've localized the problem to scale rather than spectrum.

**Spectrum, which for fire is the one I'd worry about.** S2 burn detection leans on SWIR — `swir16` at 1.61 µm and `swir22` at 2.19 µm, which is what NBR is built from. NAIP stops at `nir` 0.842 µm. Your head had ten bands including both SWIRs available in the tokens it learned from; on NAIP those wavelengths were never measured. Even with normalization and GSD perfect, a head that keyed on SWIR-driven features has nothing to key on in a four-band RGB+NIR embedding. That would show up exactly as "not getting anything at all".

If you want to separate "my preprocessing is wrong" from "the transfer doesn't hold", embed the same scene both ways and compare before involving the head at all: cosine similarity between the two CLS embeddings, and the per-channel mean and std of the patch tokens. If the NAIP token statistics sit far outside the range your head was trained on, the head is being asked to extrapolate and the mask collapsing to nothing is the expected outcome, not a bug.
